File: paint/sources/material_nodes/vector_math2_node.ts

function vector_math2_node_init() {
    array_push(nodes_material_converter, vector_math2_node_def);
    map_set(parser_material_node_vectors, "VECT_MATH", vector_math2_node_vector);
	map_set(parser_material_node_values, "VECT_MATH", vector_math2_node_value);
}

function vector_math2_node_vector(node: ui_node_t, socket: ui_node_socket_t): string {
    let vec1: string = parser_material_parse_vector_input(node.inputs[0]);
    let vec2: string = parser_material_parse_vector_input(node.inputs[1]);
    let but: ui_node_button_t = node.buttons[0]; //operation;
    let op: string = to_upper_case(u8_array_string_at(but.data, but.default_value[0]));
    op = string_replace_all(op, " ", "_");
    if (op == "ADD") {
        return "(" + vec1 + " + " + vec2 + ")";
    }
    else if (op == "SUBTRACT") {
        return "(" + vec1 + " - " + vec2 + ")";
    }
    else if (op == "AVERAGE") {
        return "((" + vec1 + " + " + vec2 + ") / 2.0)";
    }
    else if (op == "DOT_PRODUCT") {
        return parser_material_to_vec3("dot(" + vec1 + ", " + vec2 + ")");
    }
    else if (op == "LENGTH") {
        return parser_material_to_vec3("length(" + vec1 + ")");
    }
    else if (op == "DISTANCE") {
        return parser_material_to_vec3("distance(" + vec1 + ", " + vec2 + ")");
    }
    else if (op == "CROSS_PRODUCT") {
        return "cross(" + vec1 + ", " + vec2 + ")";
    }
    else if (op == "NORMALIZE") {
        return "normalize(" + vec1 + ")";
    }
    else if (op == "MULTIPLY") {
        return "(" + vec1 + " * " + vec2 + ")";
    }
    else if (op == "DIVIDE") {
        let store: string = parser_material_store_var_name(node) + "_vec2";
        parser_material_write(parser_material_kong, "var " + store + ": float3 = " + vec2 + ";");
        parser_material_write(parser_material_kong, "if (" + store + ".x == 0.0) { " + store + ".x = 0.000001; }");
        parser_material_write(parser_material_kong, "if (" + store + ".y == 0.0) { " + store + ".y = 0.000001; }");
        parser_material_write(parser_material_kong, "if (" + store + ".z == 0.0) { " + store + ".z = 0.000001; }");
        return "(" + vec1 + " / " + vec2 + ")";
    }
    else if (op == "PROJECT") {
        return "(dot(" + vec1 + ", " + vec2 + ") / dot(" + vec2 + ", " + vec2 + ") * " + vec2 + ")";
    }
    else if (op == "REFLECT") {
        return "reflect(" + vec1 + ", normalize(" + vec2 + "))";
    }
    else if (op == "SCALE") {
        return "(" + vec2 + ".x * " + vec1 + ")";
    }
    else if (op == "ABSOLUTE") {
        return "abs3(" + vec1 + ")";
    }
    else if (op == "MINIMUM") {
        return "min3(" + vec1 + ", " + vec2 + ")";
    }
    else if (op == "MAXIMUM") {
        return "max3(" + vec1 + ", " + vec2 + ")";
    }
    else if (op == "FLOOR") {
        return "floor3(" + vec1 + ")";
    }
    else if (op == "CEIL") {
        return "ceil3(" + vec1 + ")";
    }
    else if (op == "FRACTION") {
        return "frac3(" + vec1 + ")";
    }
    else if (op == "MODULO") {
        return "(" + vec1 + " % " + vec2 + ")";
    }
    else if(op == "SNAP") {
        return "(floor3(" + vec1 + " / " + vec2 + ") * " + vec2 + ")";
    }
    else if (op == "SINE") {
        return "float3(sin(" + vec1 + ".x), sin(" + vec1 + ".y), sin(" + vec1 + ".z))";
    }
    else if (op == "COSINE") {
        return "float3(cos(" + vec1 + ".x), cos(" + vec1 + ".y), cos(" + vec1 + ".z))";
    }
    else { // TANGENT
        return "float3(tan(" + vec1 + ".x), tan(" + vec1 + ".y), tan(" + vec1 + ".z))";
    }
}

function vector_math2_node_value(node: ui_node_t, socket: ui_node_socket_t): string {
    let vec1: string = parser_material_parse_vector_input(node.inputs[0]);
    let vec2: string = parser_material_parse_vector_input(node.inputs[1]);
    let but: ui_node_button_t = node.buttons[0]; //operation;
    let op: string = to_upper_case(u8_array_string_at(but.data, but.default_value[0]));
    op = string_replace_all(op, " ", "_");
    if (op == "DOT_PRODUCT") {
        return "dot(" + vec1 + ", " + vec2 + ")";
    }
    else if (op == "LENGTH") {
        return "length(" + vec1 + ")";
    }
    else if (op == "DISTANCE") {
        return "distance(" + vec1 + ", " + vec2 + ")";
    }
    else {
        return "0.0";
    }
}

let vector_math2_node_def: ui_node_t = {
    id: 0,
    name: _tr("Vector Math"),
    type: "VECT_MATH",
    x: 0,
    y: 0,
    color: 0xff62676d,
    inputs: [
        {
            id: 0,
            node_id: 0,
            name: _tr("Vector"),
            type: "VECTOR",
            color: 0xff6363c7,
            default_value: f32_array_create_xyz(0.0, 0.0, 0.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 1
        },
        {
            id: 0,
            node_id: 0,
            name: _tr("Vector"),
            type: "VECTOR",
            color: 0xff6363c7,
            default_value: f32_array_create_xyz(0.0, 0.0, 0.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 1
        }
    ],
    outputs: [
        {
            id: 0,
            node_id: 0,
            name: _tr("Vector"),
            type: "VECTOR",
            color: 0xff6363c7,
            default_value: f32_array_create_xyz(0.0, 0.0, 0.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 0
        },
        {
            id: 0,
            node_id: 0,
            name: _tr("Value"),
            type: "VALUE",
            color: 0xffa1a1a1,
            default_value: f32_array_create_x(0.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 0
        }
    ],
    buttons: [
        {
            name: _tr("operation"),
            type: "ENUM",
            output: 0,
            default_value: f32_array_create_x(0),
            data: u8_array_create_from_string(_tr("Add") + "\n" + _tr("Subtract") + "\n" + _tr("Multiply") + "\n" + _tr("Divide") + "\n" + _tr("Average") + "\n" + _tr("Cross Product") + "\n" + _tr("Project") + "\n" + _tr("Reflect") + "\n" + _tr("Dot Product") + "\n" + _tr("Distance") + "\n" + _tr("Length") + "\n" + _tr("Scale") + "\n" + _tr("Normalize") + "\n" + _tr("Absolute") + "\n" + _tr("Minimum") + "\n" + _tr("Maximum") + "\n" + _tr("Floor") + "\n" + _tr("Ceil") + "\n" + _tr("Fraction") + "\n" + _tr("Modulo") + "\n" + _tr("Snap") + "\n" + _tr("Sine") + "\n" + _tr("Cosine") + "\n" + _tr("Tangent")),
            min: 0.0,
            max: 1.0,
            precision: 100,
            height: 0
        }
    ],
    width: 0,
    flags: 0
};

File: paint/sources/material_nodes/vector_math2_node.c

#include "../global.h"

char *vector_math2_node_vector(ui_node_t *node, ui_node_socket_t *socket) {
	char             *vec1 = parser_material_parse_vector_input(node->inputs->buffer[0]);
	char             *vec2 = parser_material_parse_vector_input(node->inputs->buffer[1]);
	ui_node_button_t *but  = node->buttons->buffer[0]; // operation;
	char             *op   = to_upper_case(u8_array_string_at(but->data, but->default_value->buffer[0]));
	op                     = string_copy(string_replace_all(op, " ", "_"));
	if (string_equals(op, "ADD")) {
		return string("(%s + %s)", vec1, vec2);
	}
	else if (string_equals(op, "SUBTRACT")) {
		return string("(%s - %s)", vec1, vec2);
	}
	else if (string_equals(op, "AVERAGE")) {
		return string("((%s + %s) / 2.0)", vec1, vec2);
	}
	else if (string_equals(op, "DOT_PRODUCT")) {
		return parser_material_to_vec3(string("dot(%s, %s)", vec1, vec2));
	}
	else if (string_equals(op, "LENGTH")) {
		return parser_material_to_vec3(string("length(%s)", vec1));
	}
	else if (string_equals(op, "DISTANCE")) {
		return parser_material_to_vec3(string("distance(%s, %s)", vec1, vec2));
	}
	else if (string_equals(op, "CROSS_PRODUCT")) {
		return string("cross(%s, %s)", vec1, vec2);
	}
	else if (string_equals(op, "NORMALIZE")) {
		return string("normalize(%s)", vec1);
	}
	else if (string_equals(op, "MULTIPLY")) {
		return string("(%s * %s)", vec1, vec2);
	}
	else if (string_equals(op, "DIVIDE")) {
		char *store = string("%s_vec2", parser_material_store_var_name(node));
		parser_material_write(parser_material_kong, string("var %s: float3 = %s;", store, vec2));
		parser_material_write(parser_material_kong, string("if (%s.x == 0.0) { %s.x = 0.000001; }", store, store));
		parser_material_write(parser_material_kong, string("if (%s.y == 0.0) { %s.y = 0.000001; }", store, store));
		parser_material_write(parser_material_kong, string("if (%s.z == 0.0) { %s.z = 0.000001; }", store, store));
		return string("(%s / %s)", vec1, vec2);
	}
	else if (string_equals(op, "PROJECT")) {
		return string("(dot(%s, %s) / dot(%s, %s) * %s)", vec1, vec2, vec2, vec2, vec2);
	}
	else if (string_equals(op, "REFLECT")) {
		return string("reflect(%s, normalize(%s))", vec1, vec2);
	}
	else if (string_equals(op, "SCALE")) {
		return string("(%s.x * %s)", vec2, vec1);
	}
	else if (string_equals(op, "ABSOLUTE")) {
		return string("abs3(%s)", vec1);
	}
	else if (string_equals(op, "POWER")) {
		return string("float3(pow(%s.x, %s.x), pow(%s.y, %s.y), pow(%s.z, %s.z))", vec1, vec2, vec1, vec2, vec1, vec2);
	}
	else if (string_equals(op, "SIGN")) {
		return string("float3(sign(%s.x), sign(%s.y), sign(%s.z))", vec1, vec1, vec1);
	}
	else if (string_equals(op, "MINIMUM")) {
		return string("min3(%s, %s)", vec1, vec2);
	}
	else if (string_equals(op, "MAXIMUM")) {
		return string("max3(%s, %s)", vec1, vec2);
	}
	else if (string_equals(op, "FLOOR")) {
		return string("floor3(%s)", vec1);
	}
	else if (string_equals(op, "CEIL")) {
		return string("ceil3(%s)", vec1);
	}
	else if (string_equals(op, "FRACTION")) {
		return string("frac3(%s)", vec1);
	}
	else if (string_equals(op, "MODULO")) {
		return string("(%s %% %s)", vec1, vec2);
	}
	else if (string_equals(op, "SNAP")) {
		return string("(floor3(%s / %s) * %s)", vec1, vec2, vec2);
	}
	else if (string_equals(op, "SINE")) {
		return string("float3(sin(%s.x), sin(%s.y), sin(%s.z))", vec1, vec1, vec1);
	}
	else if (string_equals(op, "COSINE")) {
		return string("float3(cos(%s.x), cos(%s.y), cos(%s.z))", vec1, vec1, vec1);
	}
	else { // TANGENT
		return string("float3(tan(%s.x), tan(%s.y), tan(%s.z))", vec1, vec1, vec1);
	}
}

char *vector_math2_node_value(ui_node_t *node, ui_node_socket_t *socket) {
	char             *vec1 = parser_material_parse_vector_input(node->inputs->buffer[0]);
	char             *vec2 = parser_material_parse_vector_input(node->inputs->buffer[1]);
	ui_node_button_t *but  = node->buttons->buffer[0]; // operation;
	char             *op   = to_upper_case(u8_array_string_at(but->data, but->default_value->buffer[0]));
	op                     = string_copy(string_replace_all(op, " ", "_"));
	if (string_equals(op, "DOT_PRODUCT")) {
		return string("dot(%s, %s)", vec1, vec2);
	}
	else if (string_equals(op, "LENGTH")) {
		return string("length(%s)", vec1);
	}
	else if (string_equals(op, "DISTANCE")) {
		return string("distance(%s, %s)", vec1, vec2);
	}
	else {
		return "0.0";
	}
}

void vector_math2_node_init() {

	char *vector_math_operation_data =
	    string("%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s", _tr("Add"), _tr("Subtract"),
	           _tr("Multiply"), _tr("Divide"), _tr("Average"), _tr("Cross Product"), _tr("Project"), _tr("Reflect"), _tr("Dot Product"), _tr("Distance"),
	           _tr("Length"), _tr("Scale"), _tr("Normalize"), _tr("Absolute"), _tr("Power"), _tr("Sign"), _tr("Minimum"), _tr("Maximum"), _tr("Floor"),
	           _tr("Ceil"), _tr("Fraction"), _tr("Modulo"), _tr("Snap"), _tr("Sine"), _tr("Cosine"), _tr("Tangent"));
	ui_node_t *vector_math2_node_def =
	    GC_ALLOC_INIT(ui_node_t, {.id     = 0,
	                              .name   = _tr("Vector Math"),
	                              .type   = "VECT_MATH",
	                              .x      = 0,
	                              .y      = 0,
	                              .color  = 0xff62676d,
	                              .inputs = any_array_create_from_raw(
	                                  (void *[]){
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Vector"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Vector"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                  },
	                                  2),
	                              .outputs = any_array_create_from_raw(
	                                  (void *[]){
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Vector"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 0}),
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Value"),
	                                                                       .type          = "VALUE",
	                                                                       .color         = 0xffa1a1a1,
	                                                                       .default_value = f32_array_create_x(0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 0}),
	                                  },
	                                  2),
	                              .buttons = any_array_create_from_raw(
	                                  (void *[]){
	                                      GC_ALLOC_INIT(ui_node_button_t, {.name          = _tr("operation"),
	                                                                       .type          = "ENUM",
	                                                                       .output        = 0,
	                                                                       .default_value = f32_array_create_x(0),
	                                                                       .data          = u8_array_create_from_string(vector_math_operation_data),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .height        = 0}),
	                                  },
	                                  1),
	                              .width = 0,
	                              .flags = 0});

	any_array_push(nodes_material_utilities, vector_math2_node_def);
	any_map_set(parser_material_node_vectors, "VECT_MATH", vector_math2_node_vector);
	any_map_set(parser_material_node_values, "VECT_MATH", vector_math2_node_value);
}
